openlibrary/core/search.py

```diff
@@ -0,0 +1,179 @@
+    book availability information.
+
+    Retrieves a list of available editions (one per work) on Open
+    Library using the archive.org advancedsearch API. Is used in such
+    components as IABookCarousel to retrieve a list of unique
+    available books.
+
+    """
+    # Enable method to be cacheable
+    if 'env' not in web.ctx:
+        delegate.fakeload()
+
+    q = _expand_api_query(query)
+    sorts = _normalize_sorts(sorts)
+    params = _clean_params(q=q, sorts=sorts, page=page, limit=limit)
+    url = _compose_advancedsearch_url(**params)
+    response = _request(url)
+    items = response.get('docs', [])
+    item_index = _index_items_by_ocaid(items)
+    editions = [
+        Edition.canonicalize(_add_availability_to_edition(
+            edition, item_index)).dict() for edition
+        in web.ctx.site.get_many([
+            '/books/%s' % item['openlibrary_edition']
+            for item in item_index.values()
+        ])
+    ]
+
+    return {
+        'editions': editions,
+        'advancedsearch_url': url,
+        'browse_url': _compose_browsable_url(q, sorts),
+        'params': params
+    }
+
+def _normalize_sorts(sorts):
+    """_request requires sorts to be a list of valid sort options.
+    Discard invalid sorts and marshal to list.
+    """
+    VALID_SORTS = [
+        '__random', '__sort', 'addeddate', 'avg_rating', 'call_number',
+        'createdate', 'creatorSorter', 'creatorSorterRaw', 'date',
+        'downloads', 'foldoutcount', 'headerImage', 'identifier',
+        'identifierSorter', 'imagecount', 'indexdate',
+        'item_size', 'languageSorter', 'licenseurl', 'mediatype',
+        'mediatypeSorter', 'month', 'nav_order', 'num_reviews',
+        'programSorter', 'publicdate', 'reviewdate', 'stars',
+        'titleSorter', 'titleSorterRaw', 'week',  'year',
+        'loans__status__last_loan_date'
+    ]
+    if sorts:
+        # If it's a string, split and turn to list
+        if isinstance(sorts, six.string_types):
+            sorts = sorts.split(',')
+        if isinstance(sorts, list):
+            # compare against field with +asc/desc suffix rm'd
+            # e.g: date, not date+asc or date+desc
+            return [sort.replace('+', ' ') for sort in sorts
+                    if sort.split('+')[0] in VALID_SORTS]
+    return ['']
+
+def _expand_api_query(query):
+    """Expands short / more easily cacheable preset queries
+    and fixes query to ensure only text archive.org items
+    are returned having openlibrary_work
+    """
+    # Expand preset queries
+    if query in PRESET_QUERIES:
+        query = PRESET_QUERIES[query]
+
+    q = 'mediatype:texts AND !noindex:* AND openlibrary_work:(*)'
+    # Add availability if none present (e.g. borrowable only)
+    if AVAILABILITY_STATUS not in query:
+        q += ' AND %s:AVAILABLE' % AVAILABILITY_STATUS
+    return q if not query else q + " AND " + query
+
+def _clean_params(q='', sorts='', page=1, limit=MAX_IA_RESULTS):
```

Comment:
This is what gave that type error; is `sorts` supposed to be a list or a string? Type params, please.